cf_units/_udunits2_parser/__init__.py

```diff
@@ -15,10 +15,217 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with cf-units.  If not, see <http://www.gnu.org/licenses/>.
 
+import unicodedata
 
-def py_3_only():
-    # This function is only syntactically valid in Python 3.
-    # We want PY2 users to see this syntax error, rather than a
-    # significantly more complex SyntaxError in the generated parser.
+from antlr4 import InputStream, CommonTokenStream
+from antlr4.error.ErrorListener import ErrorListener
 
-    return f'is_py_3!'
+from .parser.udunits2Lexer import udunits2Lexer
+from .parser.udunits2Parser import udunits2Parser
+from .parser.udunits2ParserVisitor import udunits2ParserVisitor
+from . import graph
+
+
+# Dictionary mapping token rule id to token name.
+TOKEN_ID_NAMES = {getattr(udunits2Lexer, rule, -1): rule
+                  for rule in udunits2Lexer.ruleNames}
+
+
+def handle_UNICODE_EXPONENT(string):
+    # Convert unicode to compatibility form, replacing unicode minus with
+    # ascii minus (which is actually a less good version
+    # of unicode minus).
+    normd = unicodedata.normalize('NFKC', string).replace('−', '-')
+    return int(normd)
+
+
+class UnitParseVisitor(udunits2ParserVisitor):
+    """
+    A visitor which converts the parse tree into an abstract expression graph.
+
+    """
+    def defaultResult(self):
+        # Called once per ``visitChildren`` call.
+        return []
+
+    def aggregateResult(self, aggregate, nextResult):
+        # Always result a list from visitChildren
+        # (default behaviour is to return the last element).
+        if nextResult is not None:
+            aggregate.append(nextResult)
+        return aggregate
+
+    def visitChildren(self, node):
+        # If there is only a single item in the visitChildren's list,
+        # return the item. The list itself has no semantics.
+        result = super().visitChildren(node)
+        while isinstance(result, list) and len(result) == 1:
+            result = result[0]
+        return result
+
+    TERM_HANDLERS = {
+            'CLOSE_PAREN': None,
+            'DATE': str,
+            'DIVIDE': graph.Operand('/'),  # Drop context, such as " per ".
+            'E_POWER': str,
+            'FLOAT': graph.Number,  # Preserve precision as str.
+            'HOUR_MINUTE_SECOND': str,
+            'HOUR_MINUTE': str,
+            'ID': graph.Identifier,
+            'INT': lambda c: graph.Number(int(c)),
+            'MULTIPLY': graph.Operand('*'),
+            'OPEN_PAREN': None,
+            'PERIOD': str,
+            'RAISE': graph.Operand,
+            'TIMESTAMP': graph.Timestamp,
+            'SIGNED_INT': lambda c: graph.Number(int(c)),
+            'SHIFT_OP': None,
+            'WS': None,
+            'UNICODE_EXPONENT': handle_UNICODE_EXPONENT,
+    }
+
+    def visitTerminal(self, ctx):
+        content = ctx.getText()
+
+        symbol_idx = ctx.symbol.type
+        if symbol_idx == -1:
```

Comment:
@pelson Is this `-1` the same as the `-1` from [line +30](https://github.com/SciTools/cf-units/pull/140/files#diff-7bbcd412a73e0276cd5f54a21f54f98bR30) when constructing the `TOKEN_ID_NAMES` dictionary?

If so, should this be a first class CONSTANT of some kind?

Reply:
They aren't the same ``-1``  here, though it isn't entirely transparent what is going on. The -1 here is the special EOF token. The ``-1`` in the ``TOKEN_ID_NAMES`` is because there are non-terminals in the lexer rules (aka fragments).